src/interchange/avro.rs

```diff
@@ -653,7 +653,11 @@ impl Encoder {
                 Value::Union(1, Box::new(row))
             }
         };
-        let val = Value::Record(vec![("before".into(), before), ("after".into(), after)]);
+        Value::Record(vec![("before".into(), before), ("after".into(), after)])
+    }
+
+    fn diff_pair_to_avro(&self, diff_pair: DiffPair<&Row>) -> Vec<u8> {
```

Comment:
This method is only used in once place and is kind of suboptimal API. (Not your fault, but the Avro package's fault.) Here's an easy refactor that saves an allocation:

```diff
diff --git a/src/avro/lib.rs b/src/avro/lib.rs
index ea43d819..3c7e6617 100644
--- a/src/avro/lib.rs
+++ b/src/avro/lib.rs
@@ -335,7 +335,7 @@ pub use crate::reader::{from_avro_datum, Reader};
 pub use crate::schema::{ParseSchemaError, Schema};
 pub use crate::types::SchemaResolutionError;
 pub use crate::util::{max_allocation_bytes, DecodeError};
-pub use crate::writer::{to_avro_datum, ValidationError, Writer};
+pub use crate::writer::{to_avro_datum, write_avro_datum, ValidationError, Writer};
 
 #[cfg(test)]
 mod tests {
diff --git a/src/avro/writer.rs b/src/avro/writer.rs
index dd2a69d4..b113a12c 100644
--- a/src/avro/writer.rs
+++ b/src/avro/writer.rs
@@ -278,7 +278,7 @@ impl<W: Write> Writer<W> {
 ///
 /// This is an internal function which gets the bytes buffer where to write as parameter instead of
 /// creating a new one like `to_avro_datum`.
-fn write_avro_datum<T: ToAvro>(
+pub fn write_avro_datum<T: ToAvro>(
     schema: &Schema,
     value: T,
     buffer: &mut Vec<u8>,
diff --git a/src/dataflow/sink/avro_ocf.rs b/src/dataflow/sink/avro_ocf.rs
index 91db9fe0..be6af07b 100644
--- a/src/dataflow/sink/avro_ocf.rs
+++ b/src/dataflow/sink/avro_ocf.rs
@@ -58,7 +58,7 @@ pub fn avro_ocf<G>(
                             after: Some(row),
                         }
                     };
-                    let value = encoder.diff_pair_to_avro_value(diff_pair);
+                    let value = encoder.diff_pair_to_avro(diff_pair);
                     for _ in 0..diff.abs() {
                         let res = avro_writer.append(value.clone());
 
diff --git a/src/interchange/avro.rs b/src/interchange/avro.rs
index 53ee7cec..2a16a362 100644
--- a/src/interchange/avro.rs
+++ b/src/interchange/avro.rs
@@ -634,11 +634,16 @@ impl Encoder {
         buf.write_u8(0).expect("writing to vec cannot fail");
         buf.write_i32::<NetworkEndian>(schema_id)
             .expect("writing to vec cannot fail");
-        buf.extend(self.diff_pair_to_avro(diff_pair));
+        avro::write_avro_datum(
+            &self.writer_schema,
+            self.diff_pair_to_avro(diff_pair),
+            &mut buf,
+        )
+        .expect("schema constructed to match val");
         buf
     }
 
-    pub fn diff_pair_to_avro_value(&self, diff_pair: DiffPair<&Row>) -> Value {
+    pub fn diff_pair_to_avro(&self, diff_pair: DiffPair<&Row>) -> Value {
         let before = match diff_pair.before {
             None => Value::Union(0, Box::new(Value::Null)),
             Some(row) => {
@@ -656,11 +661,6 @@ impl Encoder {
         Value::Record(vec![("before".into(), before), ("after".into(), after)])
     }
 
-    fn diff_pair_to_avro(&self, diff_pair: DiffPair<&Row>) -> Vec<u8> {
-        let val = self.diff_pair_to_avro_value(diff_pair);
-        avro::to_avro_datum(&self.writer_schema, val).expect("schema constructed to match val")
-    }
-
     fn row_to_avro(&self, row: Vec<Datum>) -> Value {
         let fields = self
             .desc
```